cmd/terminal.go

```diff
@@ -0,0 +1,367 @@
+	}
+	defer func() {
+		_ = term.Restore(termID, oldState)
+	}()
+
+	// send the shell start message
+	m := &ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:   ws.ProtoTypeShell,
+			MsgType: wsshell.MessageTypeSpawnShell,
+			Properties: map[string]interface{}{
+				"terminal_width":  termWidth,
+				"terminal_height": termHeight,
+			},
+		},
+	}
+
+	data, err := msgpack.Marshal(m)
+	if err != nil {
+		return errors.Wrap(err, "Unable to parse the message from the websocket")
+	}
+	_ = conn.SetWriteDeadline(time.Now().Add(writeWait))
+	_ = conn.WriteMessage(websocket.BinaryMessage, data)
+
+	// message channel
+	msgChan := make(chan *ws.ProtoMsg)
+
+	c.running = true
+	go c.pipeStdout(msgChan, os.Stdout)
+	go c.pipeStdin(conn, os.Stdin)
+
+	// handle CTRL+C and signals
+	quit := make(chan os.Signal, 1)
+	signal.Notify(quit, unix.SIGINT, unix.SIGTERM)
+
+	// resize the terminal window
+	go c.resizeTerminal(ctx, msgChan, termID, termWidth, termHeight)
+
+	// wait for CTRL+C, signals or stop
+	for c.running {
+		select {
+		case msg := <-msgChan:
+			data, err := msgpack.Marshal(msg)
+			if err != nil {
+				log.Err(fmt.Sprintf("error: %v", err))
+				break
+			}
+			_ = conn.SetWriteDeadline(time.Now().Add(writeWait))
+			_ = conn.WriteMessage(websocket.BinaryMessage, data)
+		case <-ticker.C:
+			pongWaitString := strconv.Itoa(int(pongWait.Seconds()))
+			_ = conn.WriteControl(
+				websocket.PingMessage,
+				[]byte(pongWaitString),
+				time.Now().Add(writeWait),
+			)
+		case <-quit:
+			c.running = false
+		case <-c.stop:
+			c.running = false
+		}
+	}
+
+	// cancel the context
+	cancelContext()
+
+	// send the stop shell message
+	m = &ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:     ws.ProtoTypeShell,
+			MsgType:   wsshell.MessageTypeStopShell,
+			SessionID: c.sessionID,
+		},
+	}
+
+	data, err = msgpack.Marshal(m)
+	if err != nil {
+		return errors.Wrap(err, "Unable to parse the message from the websocket")
+	}
+	_ = conn.SetWriteDeadline(time.Now().Add(writeWait))
+	_ = conn.WriteMessage(websocket.BinaryMessage, data)
+	conn.Close()
+
+	return c.err
+}
+
+func (c *TerminalCmd) resizeTerminal(ctx context.Context, msgChan chan *ws.ProtoMsg, termID int, termWidth int, termHeight int) {
+	ticker := time.NewTicker(500 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			newTermWidth, newTermHeight, _ := terminal.GetSize(termID)
+			if newTermWidth != termWidth || newTermHeight != termHeight {
+				termWidth = newTermWidth
+				termHeight = newTermHeight
+				m := &ws.ProtoMsg{
+					Header: ws.ProtoHdr{
+						Proto:   ws.ProtoTypeShell,
+						MsgType: wsshell.MessageTypeResizeShell,
+						Properties: map[string]interface{}{
+							"terminal_width":  termWidth,
+							"terminal_height": termHeight,
+						},
+					},
+				}
+				msgChan <- m
+			}
+		}
+	}
+}
+
+func (c *TerminalCmd) Stop() {
+	c.stop <- true
+	c.running = false
```

Comment:
Data race and change type:
```suggestion
	c.running = false
	c.stop <- struct{}{}
```